File: server/python/src/metaobjects/codegen/generators/output_parser_generator.py
"""Response parser codegen — one ``<template_name>_response_parser.py`` per
responding ``template.prompt`` (one declaring ``@responseRef``).

FR-006 (Python) per ADR-0010 and ``docs/superpowers/specs/2026-05-25-fr6-python-template-output-parser.md``.

ADR-0052 — a template subtype's axis is DIRECTION. ``template.output`` renders
OUTBOUND (a document, an email) and gets no parser at all; the shape a reply is
parsed INTO is ``@responseRef`` on a ``template.prompt``, never ``@payloadRef``,
which types the request the prompt renders outbound. The direction rule itself
lives in :mod:`~metaobjects.codegen.generators.find_inbound` and is never
re-derived here.

Single-API throw-only convention matches the Python ecosystem norm: Pydantic
raises ``pydantic.ValidationError`` on bad input; callers wrap in ``try/except``
as needed. No dual API — TS uses ``parseX``/``safeParseX`` because Zod's
``safeParse`` is idiomatic; C# uses ``Parse``/``TryParse`` per BCL convention;
Python's ecosystem (Pydantic, Instructor, FastAPI, LangChain structured-output)
is throw-only and a dual surface would feel un-Pythonic.

Import-style emit (ADR-0056): the parser module is a thin ``parse_<name>(text) -> <Vo>``
wrapper that imports the ``@responseRef`` value object's own Pydantic model from the
module the ``entity`` generator emits (``<Vo>.py``) and declares no model of its own. The
same model is what the extract tier builds — one type per value object, in every port.
A run that wires this generator therefore needs ``entity`` too.

The strict ``parse_*`` tier is JSON-ONLY, per ADR-0053: an XML reply gets the
tolerant extract and nothing strict (see
:func:`~metaobjects.codegen.generators.find_inbound.is_xml`)."""
from __future__ import annotations

from collections.abc import Callable

from metaobjects.apidocs.naming import snake_case as _snake_case
from metaobjects.codegen import extract_delegate_emitter as rde
from metaobjects.codegen.constants import generated_header
from metaobjects.codegen.format import ruff_format
from metaobjects.codegen.generator import EmittedFile, GenContext, Generator
from metaobjects.codegen.generators.find_inbound import (
    inbound_templates,
    is_xml,
    response_shape,
)
from metaobjects.codegen.value_objects import model_class_name, model_import, pkg_of
from metaobjects.meta.core.object.meta_object import MetaObject
from metaobjects.meta.meta_data import MetaData

_GENERATOR_NAME = "output-parser-generator"


def _payload_name_collides(root: MetaData, payload: MetaObject) -> bool:
    """ADR-0044 (#228) — True iff more than one root-level ``MetaObject`` (ANY
    subtype — matching the exact domain the GENERATED runtime lookup itself
    scans: ``isinstance(child, MetaObject)``, no subtype filter) shares
    *payload*'s bare ``name``.

    The emitted ``extract_lenient_*_with_loader`` resolves its payload at
    RUNTIME via a bare-name-only, load-order-dependent first-match scan over
    ``root.own_children()`` — the same ADR-0042 "wrong node" hazard class
    ``ref_vo`` had before #228, just one layer up (a generated-code runtime
    lookup, not a build-time codegen resolution). When two responding
    ``template.prompt``s in different packages declare an own ``@responseRef``
    shape that shares a bare name, that scan could silently bind whichever object
    the loader happened to iterate first.

    Mirrors the TS reference (``output-parser.ts``'s ``payloadNameCollides``).
    Computed directly against ``root.own_children()`` — the actual domain the
    generated scan searches — rather than read off the value-object name map,
    whose domain is value objects only."""
    return (
        sum(
            1
            for c in root.own_children()
            if isinstance(c, MetaObject) and c.name == payload.name
        )
        > 1
    )


def render_output_parser(template: MetaData, root: MetaData) -> str | None:
    """Render one parser module for a responding ``template.prompt`` node.

    The emitted module imports the ``@responseRef`` value object's model from the
    entity generator's ``<Vo>.py`` and exposes a throw-only ``parse_<name>(text)``
    entry point.

    Returns ``None`` when the template declares no ``@responseRef`` or the ref
    does not resolve (defensive; the loader's template-validation pass would
    normally catch an unresolvable ref first)."""
    # ADR-0052: the shape parsed INTO is @responseRef — the reply. `response_shape`
    # resolves it through the SAME value-object target rule @payloadRef obeys, so a
    # parser can never bind a shape that has no generated model.
    shape = response_shape(root, template, pkg_of(template))
    if shape is None:
        return None
    payload = shape.vo

    # ADR-0044 (#228) — computed once, up front: does more than one root-level
    # object share this response VO's bare name? Drives BOTH the conditional
    # `resolve_object_ref` import below and the PAYLOAD_NAME/lookup emission
    # further down.
    payload_name_collides = _payload_name_collides(root, payload)

    template_name = template.name
    snake = _snake_case(template_name)
    # ADR-0056: the strict result IS the response value object's own model; the entity
    # generator writes it as `<Vo>.py`, so the module name is the class name.
    payload_class = model_class_name(payload)
    parse_fn = f"parse_{snake}"

    fqn = (
        f"{payload.package}::{template_name}"
        if payload.package
        else template_name
    )

    # ADR-0052/0053: the tolerant extract is now UNCONDITIONAL — declaring a response
    # shape IS the request for one, and @responseFormat is a closed json|xml set, so
    # there is no third case left to gate on. The reply's syntax comes from
    # @responseFormat, never @format (the syntax of the rendered prompt BODY); the old
    # @format gate is what let a text-bodied prompt asking for a JSON answer get a
    # strict parser and no extract at all.
    #
    # The strict Pydantic tier is JSON-ONLY: `model_validate_json` is an exact parser,
    # which is what makes strict all-or-nothing meaningful. Layering it over the
    # REPAIRING XML reader would raise or accept based on how much repair happened —
    # not a contract anyone can reason about. So an XML reply gets the tolerant
    # extract and nothing strict.
    emit_strict = not is_xml(shape.format)
    extracted_class = rde.mirror_name(payload)
    extract_lenient_fn = f"extract_lenient_{snake}"

    lines: list[str] = [
        generated_header(template_name, fqn),
        "from __future__ import annotations\n",
    ]

    lines.append("from dataclasses import dataclass")
    lines.append("")
    lines.append("from metaobjects.render import (")
    lines.append("    Format,")
    lines.append("    ExtractOptions,")
    lines.append("    ExtractionResult,")
    lines.append(")")
    # FR-010: the single metadata-driven extract path resolves the payload
    # MetaObject from a loaded MetaRoot and delegates to the runtime extract
    # (which assembles the FULL nested object graph reflection-free by reading
    # the live metadata directly). Codegen-wrapping-runtime — mirrors the
    # Java/Kotlin/TS pilots.
    #
    # ADR-0044 (#228) — the payload lookup below is EITHER a bare
    # `root.own_children()` scan (needs the `MetaObject` isinstance check) OR,
    # on a bare-name collision, a canonical FQN-exact `resolve_object_ref` (needs
    # no `MetaObject` import) — so exactly ONE of these two imports is emitted,
    # never both, keeping the generated module import-clean either way.
    if payload_name_collides:
        lines.append("from metaobjects.naming_refs import resolve_object_ref")
    else:
        lines.append(
            "from metaobjects.meta.core.object.meta_object import MetaObject"
        )
    lines.append(
        "from metaobjects.meta.core.object.object_extract import extract_object"
    )
    lines.append("from metaobjects.meta.meta_root import MetaRoot")
    lines.append("")

    # The model is imported ONLY by the strict tier — an XML reply's module would
    # otherwise carry an unused import (the tolerant tier's return type is the
    # all-nullable ``…Extracted`` mirror declared right here).
    if emit_strict:
        lines.append(model_import(payload))
    lines.append("")
    lines.append("")
    if emit_strict:
        lines.extend(
            [
                f"def {parse_fn}(text: str) -> {payload_class}:",
                f'    """Parse an LLM response into a typed ``{payload_class}``.',
                "",
                "    Raises:",
                "        pydantic.ValidationError: when the input does not match the schema.",
                '    """',
                f"    return {payload_class}.model_validate_json(text)",
                "",
                "",
            ]
        )

    # FR-010 nested-AWARE extracted mirror: one ``<Vo>Extracted`` dataclass for the
    # response value object and every reachable nested one, each named after its value
    # object's model (ADR-0056 rule 3). The single (delegating) extract path returns it.
    lines.extend(rde.nested_mirror_dataclasses(payload, root))
    lines.append("")
    lines.append("")

    # ---- Runtime-delegating extract (the single metadata-driven extract path) ----
    # The baked PAYLOAD_NAME is normally the resolved payload VO's SIMPLE name: the
    # delegating entry resolves the MetaObject from a loaded MetaRoot by it
    # (root child named ``payload.name``), then delegates to the runtime
    # ``extract_object`` (FULL nested graph, reflection-free) and maps the
    # assembled ValueObject graph into the typed nullable mirror graph.
    #
    # ADR-0044 (#228) — a bare ``root.own_children()`` first-match scan (below) is a
    # load-order-dependent "wrong node" hazard when this payload's OWN bare name
    # collides with another root-level object elsewhere in the run (see
    # ``_payload_name_collides``). When it does, PAYLOAD_NAME bakes the FQN
    # (``resolution_key()``) and the lookup resolves via the canonical ADR-0042
    # ``resolve_object_ref`` (FQN-exact, load-order-independent) instead of the
    # scan. A non-colliding payload keeps the bare name + the scan — byte-identical
    # to pre-#228 output.
    format_enum = "Format.XML" if is_xml(shape.format) else "Format.JSON"
    root_mapper = rde.mapper_name(payload)
    extract_lenient_with_fn = f"{extract_lenient_fn}_with_loader"
    baked_payload_name = (
        payload.resolution_key() if payload_name_collides else payload.name
    )
    lines.append("#: Payload value-object name this parser extracts — resolved")
    if payload_name_collides:
        lines.append("#: against a loaded MetaRoot at runtime. ADR-0042 FQN (this")
        lines.append("#: payload's bare name collides with a same-short-name")
        lines.append("#: object elsewhere in the run).")
    else:
        lines.append("#: against a loaded MetaRoot at runtime.")
    lines.append(f'PAYLOAD_NAME = "{baked_payload_name}"')
    lines.append("")
    lines.append("")
    lines.extend(rde.nested_mappers(payload, root))
    lines.append("")
    lines.append("")
    lines.extend(rde.delegate_helpers(rde.used_helpers(payload, root)))
    lines.append("")
    lines.append("")
    lines.append(
        f"def {extract_lenient_with_fn}("
        "root: MetaRoot, text: str, opts: ExtractOptions | None = None"
        f") -> ExtractionResult[{extracted_class}]:"
    )
    lines.append(
        '    """Runtime-delegating tolerant best-effort extraction; never raises.'
    )
    lines.append("    FULLY populates nested-object and array-of-object components by")
    lines.append("    delegating to the metadata-driven runtime ``extract_object`` (which")
    lines.append("    assembles the whole graph reflection-free via the Phase A object")
    lines.append("    model, reading the live metadata directly), then maps the assembled")
    lines.append(f"    graph into the typed ``{extracted_class}`` mirror.")
    lines.append("")
    lines.append("    :param root: a loaded ``MetaRoot`` that declares the")
    lines.append(f'                 ``{payload.name}`` value-object."""')
    if payload_name_collides:
        # ADR-0044 (#228) — PAYLOAD_NAME is a baked FQN; resolve it via the
        # canonical ADR-0042 package-local contract (FQN-exact here, so the
        # "" referrer package is inert) rather than a bare load-order-dependent
        # scan.
        lines.append('    mo = resolve_object_ref(root, PAYLOAD_NAME, "")')
    else:
        lines.append("    mo = None")
        # Emits a root-scan into generated code: root is the loader ROOT (never
        # extended, so own == effective) — ADR-0039 sanctioned own in emitted code.
        lines.append("    for child in root.own_children():")
        lines.append("        if (")
        lines.append("            isinstance(child, MetaObject)")
        lines.append("            and child.name == PAYLOAD_NAME")
        lines.append("        ):")
        lines.append("            mo = child")
        lines.append("            break")
    lines.append("    if mo is None:")
    lines.append("        raise ValueError(")
    lines.append(
        f'            f"{extract_lenient_with_fn}: payload \'{{PAYLOAD_NAME}}\' not found "'
    )
    lines.append('            "in the supplied MetaRoot"')
    lines.append("        )")
    lines.append(
        f"    outcome = extract_object(mo, text, {format_enum}, opts)"
    )
    lines.append(f"    data = {root_mapper}(outcome.data)")
    lines.append("    return ExtractionResult(data=data, report=outcome.report)")
    lines.append("")
    lines.append("")
    exported = [extract_lenient_with_fn, extracted_class, "PAYLOAD_NAME"]
    if emit_strict:
        exported.insert(0, parse_fn)
    joined = ", ".join(f'"{n}"' for n in exported)
    lines.append(f"__all__ = [{joined}]")

    lines.append("")
    return "\n".join(lines)


class OutputParserGenerator:
    """Generator wrapping ``render_output_parser``. Emits one file per responding
    ``template.prompt`` declared at root level (ADR-0052)."""

    name = _GENERATOR_NAME

    def __init__(self, *, filter: Callable[[MetaObject], bool] | None = None) -> None:
        # The ``filter`` arg matches the cross-generator contract even though
        # this generator iterates templates (not entities) and doesn't apply
        # entity-level filters today.
        self.filter = filter

    def _render_module(self, template: MetaData, root: MetaData) -> str | None:
        """EXTENSION SEAM — render the whole parser module for one responding
        ``template.prompt``. Defaults to :func:`render_output_parser` (the strict
        ``parse_*`` + the FR-010 tolerant ``extract_lenient_*`` twins). Override to
        pre/post-process the emitted source, or to replace the strict-parser /
        lenient-extractor emission entirely. Output is byte-identical to the default
        when not overridden."""
        return render_output_parser(template, root)

    def generate(self, ctx: GenContext) -> list[EmittedFile]:
        root = ctx.loaded_root
        if root is None:
            return []
        files: list[EmittedFile] = []
        # ADR-0052: the direction rule lives in FindInbound, never re-derived here.
        # Only a RESPONDING template.prompt gets a parser file; template.output
        # renders outbound and parses nothing.
        for tmpl in inbound_templates(root):
            content = self._render_module(tmpl, root)
            if content is None:
                ctx.warn(
                    f"{_GENERATOR_NAME}: skipping template.prompt "
                    f"'{tmpl.name}' (@responseRef does not resolve to a payload target)."
                )
                continue
            files.append(
                EmittedFile(
                    path=f"{_snake_case(tmpl.name)}_response_parser.py",
                    content=ruff_format(content),
                )
            )
        return files


def output_parser_generator(
    *, filter: Callable[[MetaObject], bool] | None = None
) -> Generator:
    """Factory mirroring the TS ``outputParser()`` and C# ``OutputParserGenerator``."""
    return OutputParserGenerator(filter=filter)

File: server/csharp/MetaObjects.Codegen/Generators/OutputParserGenerator.cs
// output-parser-generator — for each responding `template.prompt` (one carrying
// `@responseRef`), emits a `<PromptName>.response.cs` file declaring a static
// `<PromptName>Parser` class with the .NET BCL `Parse`/`TryParse` dual API
// (ADR-0010). `Parse` throws on bad input; `TryParse` returns a bool and an
// out-error.
//
// ADR-0052 — this tier is INBOUND: it reads a model's REPLY. It used to key on
// `template.output` with no format filter at all, so an email template generated
// a JSON parser for text the system had just rendered. `template.output` is
// outbound only and emits nothing here.
//
// ADR-0056: the emitted parser RETURNS the `@responseRef` value object's own POCO,
// the one EntityGenerator emits (never the `@payloadRef` object, which types the
// REQUEST this prompt renders outbound). It declares no type for the shape, so a
// run that wires it must also wire EntityGenerator (`entity`). The one type it
// does own is the value object's all-nullable mirror (`<Vo>Extracted`), and that
// is keyed by the value object: emitted ONCE per run, in the value object's own
// namespace, however many prompts parse into it.
//
// Ported from typescript/packages/codegen-ts/src/generators/output-parser-file.ts
// (renderer in typescript/packages/codegen-ts/src/templates/output-parser.ts).
// TS uses Zod; C# uses System.Text.Json. Presence is enforced by a generated
// JsonTypeInfo modifier that marks each @required field required (the POCO's
// members are shared with the REST tier, so they cannot carry the `required`
// keyword themselves), and enums travel as their member symbols.
//
// FR-010 tolerant extraction — one metadata-driven path (Plan 2.1). Every
// responding prompt also emits tolerant best-effort extraction (never throws;
// lost/malformed components are null in a nullable mirror + classified in the
// report). ADR-0052: unconditional — a declared `@responseRef` IS the request
// for the tolerant path, and `@responseFormat` is a closed json|xml set, so
// there is no third case left to gate on:
//   • Runtime-delegating  ExtractLenient(MetaObject, string[, ExtractOptions]) (+ an
//     ExtractLenient(MetaRoot, ...) convenience overload that resolves the baked
//     PAYLOAD_FQN) — delegates to MetaObjects.Codegen.Runtime.ExtractObject,
//     which assembles the FULL object graph (nested objects + arrays-of-objects)
//     reflection-free via the Phase A object model by reading the live metadata
//     directly, then maps the assembled ValueObject graph into the typed nullable
//     mirror via generated From*Extracted mappers.
//
// ASSEMBLY CONTRACT (delegating extract). ExtractObject is sited in
// MetaObjects.Codegen (it bridges core + Render, which neither can host alone).
// So the GENERATED parser, when compiled in a CONSUMER assembly, references
// MetaObjects.Codegen (ExtractObject) + MetaObjects (core: MetaObject / MetaRoot
// / ValueObject) + MetaObjects.Render (the extract engine). A consumer using the
// extract API must carry MetaObjects.Codegen on its classpath (it transitively
// brings core + Render).

using System.Text;
using MetaObjects.Meta;
using static MetaObjects.Shared.BaseTypes;
using static MetaObjects.Template.TemplateConstants;

namespace MetaObjects.Codegen.Generators;

/// <summary>
/// Emits one response-parser file per responding <c>template.prompt</c>, with a static class
/// exposing <c>Parse(string)</c> + <c>TryParse(string, out T?, out string?)</c>.
/// </summary>
public class OutputParserGenerator : IGenerator
{
    public virtual string Name => "output-parser-generator";

    public virtual IEnumerable<EmittedFile> Generate(GenContext ctx)
    {
        var files = new List<EmittedFile>();
        // ADR-0056 rule 3: each mirrored value object's <Vo>Extracted record is emitted ONCE per
        // run, in its own namespace — so two prompts parsing into one shape share it, and where it
        // lands never depends on which prompt came first.
        var mirrored = new Dictionary<string, MetaData>(StringComparer.Ordinal);
        // ADR-0052: the direction rule lives in FindInbound, never re-derived here.
        foreach (var tmpl in FindInbound.InboundTemplates(ctx.Root))
        {
            // Skip on RESOLUTION, not just presence. InboundTemplates filters on @responseRef
            // presence (it has no root to resolve against), so this loop is the only place the
            // ref's target is checked — and it was not checking. A @responseRef naming an
            // object.entity therefore emitted `static Answer Parse(string)` naming a type no
            // value-object generator declares: CS0246. Every sibling inbound
            // generator already skipped on ResponseShape; this one alone did not.
            if (FindInbound.ResponseShape(ctx.Root, tmpl) is not { } shape) continue;
            files.Add(EmitParser(tmpl, shape.Ref, ctx));
            foreach (var vo in ExtractDelegateEmitter.MirrorClosure(shape.Vo, ctx.Root))
                mirrored.TryAdd(vo.ResolutionKey(), vo);
        }
        foreach (var vo in mirrored.Values)
            files.Add(ExtractDelegateEmitter.MirrorFile(vo, ctx.Root, ctx.Config));
        return files;
    }

    /// <summary>
    /// True iff this generator emits a response parser for <paramref name="tmpl"/>: a
    /// <c>template.prompt</c> carrying a <c>@responseRef</c>. Single source of truth shared by
    /// the generator loop AND the api-docs builder.
    /// </summary>
    public static bool AppliesTo(MetaData tmpl, MetaRoot root) => FindInbound.IsInbound(tmpl, root);

    protected virtual EmittedFile EmitParser(MetaData tmpl, string responseRef, GenContext ctx)
    {
        var templateName = tmpl.Name;
        var parserClass = CSharpNaming.ParserClassName(templateName);
        // ADR-0052: the shape parsed INTO is @responseRef — the reply — never @payloadRef,
        // which types the request this prompt renders outbound.
        var payloadRef = responseRef;

        // ADR-0042/#228: resolve @payloadRef package-aware — never a bare-tail/global-scan
        // fallback that could bind the WRONG package's same-short-named object under a
        // cross-package collision (the #219/#244 "wrong node" class). The loader validates
        // @payloadRef through this SAME canonical resolver; codegen must never silently walk
        // a DIFFERENT node than what was validated.
        var referrerPkg = global::MetaObjects.NamingRefs.EffectivePackage(tmpl);
        var vo = global::MetaObjects.NamingRefs.ResolveObjectRef(ctx.Root, payloadRef, referrerPkg);
        // ADR-0056: the parsed type IS the value object's own POCO, named the way every generator
        // names it (never the raw ref string, which may be an FQN).
        var ns = ctx.Config.Namespace;
        var payloadType = vo is not null
            ? ValueObjectNames.TypeRef(vo, ctx.Root, ctx.Config, ns)
            : CSharpNaming.StripPkg(payloadRef);

        // ADR-0053: the reply's syntax is @responseFormat (json|xml, default json) — never
        // @format, which is the syntax of the rendered prompt BODY. The old @format gate is
        // what made a text-bodied prompt with a JSON reply emit a strict parser and no
        // extract at all.
        var format = FindInbound.ResponseFormatOf(tmpl);
        bool emitExtract = vo is not null;

        // The strict Parse/TryParse tier is JSON-ONLY, by construction.
        //
        // Its body is `JsonSerializer.Deserialize<T>`, and there is no XML equivalent worth
        // generating. Not because no XML reader exists — MetaObjects.Render ships a forgiving
        // one — but because strict all-or-nothing semantics layered over a REPAIRING parser is
        // incoherent: it would throw or accept based on how much repair happened, which is not
        // a contract anyone can reason about. Before ADR-0052 an XML template got
        // `JsonSerializer.Deserialize` anyway — a generated method that could never work.
        //
        // So an XML reply gets the tolerant extract and nothing strict. Its typed shape is
        // `<Name>Extracted` — a nullable mirror, the honest type for a best-effort parse.
        bool emitStrict = !FindInbound.IsXml(format);

        var sb = new StringBuilder();
        sb.AppendLine("// <auto-generated/>");
        sb.AppendLine("// Generated by MetaObjects output-parser-generator. Do not edit by hand.");
        sb.AppendLine("#nullable enable");
        sb.AppendLine("using System;");
        sb.AppendLine("using System.Collections.Generic;");
        if (emitStrict) sb.AppendLine("using System.Diagnostics.CodeAnalysis;");
        if (emitStrict) sb.AppendLine("using System.Text.Json;");
        if (emitStrict) sb.AppendLine("using System.Text.Json.Serialization;");
        var required = emitStrict && vo is not null ? RequiredFields(vo, ctx) : [];
        if (required.Count > 0) sb.AppendLine("using System.Text.Json.Serialization.Metadata;");
        if (emitExtract) sb.AppendLine("using MetaObjects.Render.Extract;");
        sb.AppendLine();
        sb.AppendLine($"namespace {ns};");
        sb.AppendLine();
        sb.AppendLine($"/// <summary>Parser for LLM responses matching the <c>{templateName}</c> template.prompt.</summary>");
        sb.AppendLine($"public static class {parserClass}");
        sb.AppendLine("{");
        if (emitStrict)
        {
            sb.AppendLine("    private static readonly JsonSerializerOptions Options = new()");
            sb.AppendLine("    {");
            // Case-sensitive (default) — every value-object POCO member carries its metadata
            // field name in [JsonPropertyName], which is the key the model is asked to produce.
            sb.AppendLine("        PropertyNameCaseInsensitive = false,");
            // A field.enum travels as its member symbol, as the output-format prompt says.
            sb.AppendLine("        Converters = { new JsonStringEnumConverter() },");
            if (required.Count > 0)
                sb.AppendLine("        TypeInfoResolver = new DefaultJsonTypeInfoResolver { Modifiers = { RequireDeclaredFields } },");
            sb.AppendLine("    };");
            sb.AppendLine();
            if (required.Count > 0)
            {
                sb.AppendLine("    /// <summary>A reply that omits a <c>@required</c> field fails to parse. The value-object POCOs");
                sb.AppendLine("    /// cannot say so themselves: their members are shared with the REST tier, where presence is a");
                sb.AppendLine("    /// validation concern.</summary>");
                sb.AppendLine("    private static void RequireDeclaredFields(JsonTypeInfo info)");
                sb.AppendLine("    {");
                sb.AppendLine("        string[]? required =");
                foreach (var (type, names) in required)
                    sb.AppendLine($"            info.Type == typeof({type}) ? new[] {{ {string.Join(", ", names.Select(n => $"\"{n}\""))} }} :");
                sb.AppendLine("            null;");
                sb.AppendLine("        if (required is null) return;");
                sb.AppendLine("        foreach (var p in info.Properties)");
                sb.AppendLine("            if (Array.IndexOf(required, p.Name) >= 0) p.IsRequired = true;");
                sb.AppendLine("    }");
                sb.AppendLine();
            }
            sb.AppendLine($"    /// <summary>Parse an LLM response into a typed <see cref=\"{payloadType}\"/>.</summary>");
            sb.AppendLine($"    /// <param name=\"text\">The raw response text — expected to be JSON.</param>");
            sb.AppendLine("    /// <returns>The deserialized, validated response.</returns>");
            sb.AppendLine("    /// <exception cref=\"JsonException\">JSON is malformed, missing a <c>required</c> property, or has a type mismatch.</exception>");
            sb.AppendLine($"    public static {payloadType} Parse(string text) =>");
            sb.AppendLine($"        JsonSerializer.Deserialize<{payloadType}>(text, Options)");
            sb.AppendLine($"            ?? throw new JsonException(\"deserialized to null\");");
            sb.AppendLine();
            sb.AppendLine($"    /// <summary>Parse with explicit error handling — does not throw on validation failure.</summary>");
            sb.AppendLine("    /// <param name=\"text\">The raw response text — expected to be JSON.</param>");
            sb.AppendLine("    /// <param name=\"value\">On success, the deserialized response; otherwise null.</param>");
            sb.AppendLine("    /// <param name=\"error\">On failure, a human-readable error message; otherwise null.</param>");
            sb.AppendLine("    /// <returns><c>true</c> on success; <c>false</c> on validation failure.</returns>");
            sb.AppendLine($"    public static bool TryParse(string text, [NotNullWhen(true)] out {payloadType}? value, [NotNullWhen(false)] out string? error)");
            sb.AppendLine("    {");
            sb.AppendLine("        try");
            sb.AppendLine("        {");
            sb.AppendLine("            value = Parse(text);");
            sb.AppendLine("            error = null;");
            sb.AppendLine("            return true;");
            sb.AppendLine("        }");
            sb.AppendLine("        catch (JsonException ex)");
            sb.AppendLine("        {");
            sb.AppendLine("            value = null;");
            sb.AppendLine("            error = ex.Message;");
            sb.AppendLine("            return false;");
            sb.AppendLine("        }");
            sb.AppendLine("    }");
        }

        // FR-010: emit the single metadata-driven extract path — the runtime-DELEGATING
        // ExtractLenient(MetaObject/MetaRoot, text) overloads (delegating to
        // MetaObjects.Codegen.Runtime.ExtractObject, which assembles the full object graph
        // reflection-free by reading the live metadata directly) and the mappers into the
        // value objects' mirrors. The mirror records themselves are emitted once per run
        // (see Generate). No baked ExtractSchema snapshot.
        string formatEnum = FindInbound.IsXml(format) ? "Format.Xml" : "Format.Json";

        if (emitExtract)
            sb.Append(ExtractDelegateEmitter.DelegatingMembers(vo!, ctx.Root, ctx.Config, ns, formatEnum));

        sb.AppendLine("}");

        // ADR-0052 D4: the artifact name follows the DIRECTION axis. A parser file named
        // `.output.cs` generated from a prompt reproduces the confusion being removed.
        return new EmittedFile($"{templateName}.response.cs", sb.ToString());
    }

    /// <summary>
    /// The <c>@required</c> fields of every value object the strict parse decodes — the response
    /// value object and each one nested under it — as (type reference, JSON names) pairs. Value
    /// objects with no required field are omitted.
    /// </summary>
    private static List<(string Type, List<string> Names)> RequiredFields(MetaData vo, GenContext ctx) =>
        ExtractDelegateEmitter.MirrorClosure(vo, ctx.Root)
            .Select(v => (
                Type: ValueObjectNames.TypeRef(v, ctx.Root, ctx.Config, ctx.Config.Namespace),
                Names: Fr010FieldMapping.Fields(v).Where(Fr010FieldMapping.IsRequired).Select(f => f.Name).ToList()))
            .Where(r => r.Names.Count > 0)
            .ToList();
}
